cli/CommandLineParser.php

```diff
@@ -0,0 +1,277 @@
+				$output->errors[$name] = 'invalid input: ' . $option->getLongAlias() . ' cannot be empty';
+			}
+		}
+
+		foreach ($this->inputs as $name => $input) {
+			foreach ($input['values'] ?? $input['defaultInput'] ?? [] as $value) {
+				switch ($this->options[$name]->getTypes()['type']) {
+					case 'int':
+						if (!ctype_digit($value)) {
+							$output->errors[$name] = 'invalid input: ' . $input['aliasUsed'] . ' must be an integer';
+						}
+						break;
+					case 'bool':
+						if (filter_var($value, FILTER_VALIDATE_BOOL, FILTER_NULL_ON_FAILURE) === null) {
+							$output->errors[$name] = 'invalid input: ' . $input['aliasUsed'] . ' must be a boolean';
+						}
+						break;
+				}
+			}
+		}
+
+		return $output;
+	}
+
+	/**
+	 * @template T
+	 * @param T $output
+	 * @return T
+	 */
+	private function appendTypedValidValues($output) {
+		foreach ($this->inputs as $name => $input) {
+			$values = $input['values'] ?? $input['defaultInput'] ?? null;
+			$types = $this->options[$name]->getTypes();
+			if ($values) {
+				$validValues = [];
+				$typedValues = [];
+
+				switch ($types['type']) {
+					case 'string':
+						$typedValues = $values;
+						break;
+					case 'int':
+						$validValues = array_filter($values, static fn($value) => ctype_digit($value));
+						$typedValues = array_map(static fn($value) => (int) $value, $validValues);
+						break;
+					case 'bool':
+						$validValues = array_filter($values, static fn($value) => filter_var($value, FILTER_VALIDATE_BOOL, FILTER_NULL_ON_FAILURE) !== null);
+						$typedValues = array_map(static fn($value) => (bool) filter_var($value, FILTER_VALIDATE_BOOL, FILTER_NULL_ON_FAILURE), $validValues);
+						break;
+				}
+
+				/** @var stdClass $output */
+				$output->$name = $types['isArray'] ? $typedValues : array_pop($typedValues);
+			}
+		}
+
+		/** @var T $output */
+		return $output;
+	}
+
+	/** @param array<string,string|false>|false $getoptOutput */
+	private function getoptOutputTransformer(false|array $getoptOutput): void {
+		$getoptOutput = is_array($getoptOutput) ? $getoptOutput : [];
+
+		foreach ($getoptOutput as $alias => $value) {
+			foreach ($this->options as $name => $data) {
+				if (in_array($alias, $data->getAliases(), true)) {
+					$this->inputs[$name]['aliasUsed'] = $alias;
+					$this->inputs[$name]['values'] = $value === false
+						? [$data->getOptionalValueDefault()]
+						: (is_array($value)
+							? $value
+							: [$value]);
+				}
+			}
+		}
+	}
+
+	/**
+	 * @param array<string> $userInputs
+	 * @return array<string>
+	 */
+	private function getAliasesUsed(array $userInputs, string $regex): array {
+		$foundAliases = [];
+
+		foreach ($userInputs as $input) {
+			preg_match($regex, $input, $matches);
+
+			if(!empty($matches['short'])) {
+				$foundAliases = array_merge($foundAliases, str_split($matches['short']));
+			}
+			if(!empty($matches['long'])) {
+				$foundAliases[] = $matches['long'];
+			}
+		}
```

Comment:
```suggestion
		foreach ($userInputs as $input) {
			preg_match($regex, $input, $matches);
			if (array_key_exists('short', $matches) === false && $matches['short'] !== null) {
				$foundAliases = array_merge($foundAliases, str_split($matches['short']));
			}
			if (array_key_exists('long', $matches) === false && $matches['long'] !== null) {
				$foundAliases[] = $matches['long'];
			}
		}
```
empty()

[empty()](https://www.php.net/manual/fr/function.empty.php) do a lot of things, depending on the type of the parameter.

You always have a better function to do it: * count() for array * > 0 or !== 0 for int * strlen() for string

Reply:
I think `empty` is the better variant here, both in terms of functionality and readability. The suggested alternative is harder to read, more error prove, and incidently slower.

Reply:
My version is more verbose, as it is more explicit about what the "if ( )" should test for.
empty() 

empty() does many things, depending on the type of parameter. You always have a better function to do it: * count() for array * > 0 or !== 0 for int * strlen() for string.

The following values are considered empty:

"" (an empty string)
0 (0 as an integer)
0.0 (0 as a float)
"0" (0 as a string)
NULL
FALSE
array() (an empty array)

In this case, the "if" refers to an array(), but you have to read the rest of the code to understand it. You might as well have wanted to know if a string was not empty. With my version, you're sure of what you're looking for and what you're expecting.

Translated with DeepL.com (free version)

Reply:
I'd say these are all good points. I think `empty()` is more readable in this case but, as you point out, it can leave ambiguity as to what we are checking for and being more explicit about that is almost always a good thing.

I can't say I really favour solution one or the other here, so I'm happy to go along with whatever the consensus is.